### PR TITLE
Security Level Datums

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

Converts security levels into datums and expands the existing subsystem by a massive amount, offloads all of the global procs to this subsystem too.

You can now add new security levels with ease and without having to do this:
image
You can now do this:
image
This also enables you to give the security level a sound when it is selected, which it previously did not do.

### How to add a new security level:
1. Add a new security level define accordingly
2. Make a new security datum and fill it out accordingly

Completed testing flow:

- [x] Use comms console to set all available levels (Checking if the level is set correctly and that you can go up and down levels)
- [x] Call shuttle and use comms console to set all available levels(Checking if the shuttle mod timer changes correctly)
- [x] Check delta alert(Checking if delta alert works)
- [x] Trigger nuke(Checking alternative set security level methods work)
- [x] Spawn narsie(Ditto)

## Why It's Good For The Game

Adding security levels to this game is an absolute nightmare and I'd like to change that. As a result, maybe some people can add new levels too.

It's less old code and more modular I guess.

## Changelog

:cl:
refactor: Security level code has been refactored, please report any abnormalities to the github.
/:cl:

### CODE REVIEW

File: code/__DEFINES/subsystems.dm [hunk 1]
```diff
@@ -119,6 +119,7 @@
 #define INIT_ORDER_INSTRUMENTS 82
 #define INIT_ORDER_GREYSCALE 81
 #define INIT_ORDER_VIS 80
+#define INIT_ORDER_SECURITY_LEVEL 79
```
Comment (Mothblocks) [inline code/__DEFINES/subsystems.dm:122]: Why does this need its own init order?

Comment (Mothblocks) [inline code/__DEFINES/subsystems.dm:122]: Why does this needs its own init order?

File: code/controllers/subsystem/security_level.dm [hunk 2]
```diff
@@ -1,17 +1,109 @@
...
+	if(selected_level.sound)
+		sound_to_playing_players(selected_level.sound)
+
+/**
+ * Returns the current security level
+ *
+ * Arguments:
+ * * as_number Whether to return the security level as a text or as a number
+ */
+/datum/controller/subsystem/security_level/proc/get_current_level(as_number)
```
Comment (Mothblocks) [inline code/controllers/subsystem/security_level.dm:83]: Make these separate procs instead.

File: code/controllers/subsystem/security_level.dm [hunk 3]
```diff
@@ -1,17 +1,109 @@
...
+	else // Going down
+		minor_announce(selected_level.lowering_to_announcement, "Attention! Security level lowered to [selected_level.name]:")
+	if(selected_level.sound)
+		sound_to_playing_players(selected_level.sound)
+
+/**
+ * Returns the current security level
+ *
+ * Arguments:
+ * * as_number Whether to return the security level as a text or as a number
```
Comment (Mothblocks) [inline code/controllers/subsystem/security_level.dm:81]: Your formatting on all of these is wrong, it should look like:

```suggestion
 * * as_number - Whether to return the security level as a text or as a number
```

File: code/modules/security_levels/security_level_datums.dm [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	/// Our announcement when elevating to this level
+	var/elevating_to_announcemnt
+	/// Our configuration key for lowering to text, if set, will override the default lowering to announcement.
+	var/lowering_to_configuration_key
+	/// Our configuration key for elevating to text, if set, will override the default elevating to announcement.
+	var/elevating_to_configuration_key
+
+/datum/security_level/New()
+	. = ..()
+	if(lowering_to_configuration_key) // I'm not sure about you, but isn't there an easier way to do this?
```
Comment (Mothblocks) [inline code/modules/security_levels/security_level_datums.dm:33]: To do what?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/security_level.dm (hunk 2 above)
Comment (Gandalf2k15) [reply code/controllers/subsystem/security_level.dm:83]: Sure

File: code/controllers/subsystem/security_level.dm (hunk 3 above)
Comment (Gandalf2k15) [reply code/controllers/subsystem/security_level.dm:81]: Got it

File: code/modules/unit_tests/security_levels.dm [hunk 5]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Security Level Unit Test
+ *
+ * This test is here to ensure there are no security levels with the same name or number level. Having the same name or number level will cause problems.
+ */
+/datum/unit_test/security_levels
```
Comment (Mothblocks) [inline code/modules/unit_tests/security_levels.dm:6]: You don't include this in `_unit_tests.dm`

File: code/modules/unit_tests/security_levels.dm [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+/datum/unit_test/security_levels/Run()
+	var/list/comparison_one = subtypesof(/datum/security_level)
+	var/list/comparison_two = LAZYCOPY(comparison_one)
```
Comment (Mothblocks) [inline code/modules/unit_tests/security_levels.dm:10]: Why are you copying this at all? You don't mutate `comparison_one`. Just use it directly.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/unit_tests/security_levels.dm (hunk 6 above)
Comment (Gandalf2k15) [reply code/modules/unit_tests/security_levels.dm:10]: I think I planned on doing sometbing with it, I'll change it up

**Comment by Gandalf2k15** [comment]
Mkay should be good to go and merge

**Pull request merged**